engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java

```diff
@@ -234,8 +230,11 @@ private void updateJob(final long key, final JobRecord updatedValue, final State
       makeJobActivatable(type, key);
     }
 
-    if (deadline > 0) {
-      removeJobDeadline(deadline);
+    if (newState != State.ACTIVATED) {
+      // This only works because none of the events actually remove the deadline from the job record.
+      // If, say on job failure, the deadline is removed or reset to 0, then we would need to look at
+      // the current state of the job to determine what deadline to remove.
+      removeJobDeadline(key, updatedValue.getDeadline());
```

Comment:
:question: I'm not sure about this assumption: "none of the events actually remove the deadline from the job record"